src/index.ts

```diff
@@ -56,3 +71,89 @@ export function assertGoError<E extends Error>(result: GoResult<any, E>): assert
     throw new Error('Assertion failed. Expected error, but no error was thrown');
   }
 }
+
+// Adapted from:
+// https://github.com/then/is-promise
+export function isPromise(obj: any) {
+  return !!obj && (typeof obj === 'object' || typeof obj === 'function') && typeof obj.then === 'function';
+}
+
+export interface PromiseOptions {
+  readonly retries?: number;
+  readonly retryDelayMs?: number;
+  readonly timeoutMs?: number;
+}
+
+export interface RetryOptions extends PromiseOptions {
+  readonly retries: number;
+}
+
+export async function retryOperation<T>(operation: () => Promise<T>, options: RetryOptions): Promise<T> {
+  // We may want to use some of these options in the future
+  const attemptOptions: AttemptOptions<any> = {
+    delay: options.retryDelayMs || DEFAULT_RETRY_DELAY_MS,
+    maxAttempts: options.retries + 1,
+    initialDelay: 0,
+    minDelay: 0,
+    maxDelay: 0,
+    factor: 0,
+    timeout: options.timeoutMs || 0,
+    jitter: false,
+    handleError: null,
+    handleTimeout: null,
+    beforeAttempt: null,
+    calculateDelay: null,
+  };
+  return retry((_context) => operation(), attemptOptions);
+}
+
+export interface ContinuousRetryOptions {
+  readonly delay?: number;
+}
+
+export function promiseTimeout<T>(ms: number, promise: Promise<T>): Promise<T> {
+  let mutableTimeoutId: NodeJS.Timeout;
+  const timeout = new Promise((_res, reject) => {
+    mutableTimeoutId = setTimeout(() => {
+      reject(new Error(`Operation timed out in ${ms} ms.`));
+    }, ms);
+  });
+
+  const wrappedPromise = promise.finally(() => {
+    if (mutableTimeoutId) {
+      clearTimeout(mutableTimeoutId);
+    }
+  });
+
+  return Promise.race([wrappedPromise, timeout]) as Promise<T>;
+}
+
+export function retryOnTimeout<T>(maxTimeoutMs: number, operation: () => Promise<T>, options?: ContinuousRetryOptions) {
+  const promise = new Promise<T>((resolve, reject) => {
+    function run(): Promise<any> {
+      // If the promise is successful, resolve it and bubble the result up
+      return operation()
+        .then(resolve)
+        .catch((reason: any) => {
+          // Only if the error is a timeout error, do we retry the promise
+          if (reason instanceof Error && reason.message.includes('Operation timed out')) {
+            // Delay the new attempt slightly
+            return sleep(options?.delay || DEFAULT_RETRY_DELAY_MS)
+              .then(run)
+              .then(resolve)
+              .catch(reject);
+          }
+
+          // If the error is NOT a timeout error, then we reject immediately
+          return reject(reason);
+        });
+    }
+
+    return run();
+  });
+
+  return promiseTimeout(maxTimeoutMs, promise);
+}
+
+export const retryGo = <T>(fn: () => Promise<T>, options?: PromiseOptions) =>
+  go(() => retryOnTimeout(DEFAULT_RETRY_TIMEOUT_MS, fn), options);
```

Comment:
I think we should provide 4 functions:
* `go` - 0 retries, no timeout
* `retryGo` - same as `go` with a default number of retries, let's say 3
* `timeoutGo` - same as `go` with a default timeout, let's say 10s
* `retryTimeoutGo` - same as `go` with a default number of retries (3) and a default timeout (10s)

And you can pass the options object to all of them to override the defaults.